utils/zerovec/src/zerovec/mod.rs

```diff
@@ -87,36 +87,85 @@ pub struct ZeroVec<'a, T>
 where
     T: AsULE,
 {
-    /// Pointer to data
-    buf: *mut [T::ULE],
-    /// Borrowed if zero. Capacity of buffer above if not
-    capacity: usize,
+    vector: EyepatchHackVector<T::ULE>,
+
     /// Marker type, signalling variance and dropck behavior
     /// by containing all potential types this type represents
     #[allow(clippy::type_complexity)] // needed to get correct marker type behavior
     marker: PhantomData<(Vec<T::ULE>, &'a [T::ULE])>,
 }
 
+// Send inherits as long as all fields are Send, but also references are Send only
+// when their contents are Sync (this is the core purpose of Sync), so
+// we need a Send+Sync bound since this struct can logically be a vector or a slice.
 unsafe impl<'a, T: AsULE> Send for ZeroVec<'a, T> where T::ULE: Send + Sync {}
+// Sync typically inherits as long as all fields are Sync
 unsafe impl<'a, T: AsULE> Sync for ZeroVec<'a, T> where T::ULE: Sync {}
 
 impl<'a, T: AsULE> Deref for ZeroVec<'a, T> {
     type Target = ZeroSlice<T>;
     #[inline]
     fn deref(&self) -> &Self::Target {
-        let slice = unsafe { &*self.buf };
+        //
+        let slice: &[T::ULE] = self.vector.as_slice();
         ZeroSlice::from_ule_slice(slice)
     }
 }
 
+// Represents an unsafe potentially-owned vector/slice type, without a lifetime
+// working around dropck limitations.
+//
+// Must either be constructed by deconstructing a Vec<U>, or from &[U] with capacity set to
+// zero. Should not outlive its source &[U] in the borrowed case; this type does not in
+// and of itself uphold this guarantee, but the .as_slice() method assumes it.
+//
+// After https://github.com/rust-lang/rust/issues/34761 stabilizes,
+// we should remove this type and use #[may_dangle]
```

Comment:
Suggestion (optional): add a canary test?

I think this solution is just fine, though